File: sdk/a2a_settlement/agentcard.py
from __future__ import annotations

from typing import Any

from a2a_settlement import A2A_SE_EXTENSION_URI


def build_settlement_extension(
    *,
    exchange_urls: list[str] | str,
    account_ids: dict[str, str] | str,
    preferred_exchange: str | None = None,
    pricing: dict[str, Any] | None = None,
    currency: str = "ATE",
    reputation: float | None = None,
    availability: float | None = None,
    required: bool = False,
    description: str = "Accepts token-based payment via A2A Settlement Exchange",
) -> dict[str, Any]:
    """Build an AgentCard `capabilities.extensions[]` entry for A2A-SE.

    Supports both v0.5 multi-exchange format and v0.2 single-exchange format.
    Returns a plain dict so callers can use it with or without `a2a-sdk` installed.

    Args:
        exchange_urls: Single URL string or list of exchange URLs.
        account_ids: Dict mapping exchange URLs to account IDs, or a single account ID string.
        preferred_exchange: Preferred exchange URL (defaults to first in list).
        pricing: Pricing config keyed by skill ID.
        currency: Default currency (default: "ATE").
        reputation: Agent's reputation score (0.0-1.0).
        availability: Agent's availability score (0.0-1.0).
        required: Whether settlement is required for this agent.
        description: Human-readable description for the extension.
    """
    if isinstance(exchange_urls, str):
        exchange_urls = [exchange_urls]
    if isinstance(account_ids, str):
        account_ids = {exchange_urls[0]: account_ids}

    params: dict[str, Any] = {
        "exchangeUrls": exchange_urls,
        "preferredExchange": preferred_exchange or exchange_urls[0],
        "accountIds": account_ids,
        "currency": currency,
    }
    if pricing is not None:
        params["pricing"] = pricing
    if reputation is not None:
        params["reputation"] = float(reputation)
    if availability is not None:
        params["availability"] = float(availability)

    return {
        "uri": A2A_SE_EXTENSION_URI,
        "description": description,
        "required": bool(required),
        "params": params,
    }

File: sdk-ts/src/agentcard.ts
export const A2A_SE_EXTENSION_URI =
  "https://a2a-settlement.org/extensions/settlement/v1";

export interface PricingEntry {
  baseTokens: number;
  model: "per-request" | "per-unit" | "per-minute" | "negotiable";
  currency?: string;
  unitSize?: number;
}

export interface SettlementExtensionOptions {
  /** Exchange URLs the agent is registered on. */
  exchangeUrls: string[];
  /** Map of exchange URL -> account ID. */
  accountIds: Record<string, string>;
  /** Preferred exchange URL (defaults to first in list). */
  preferredExchange?: string;
  /** Per-skill pricing configuration. */
  pricing?: Record<string, PricingEntry>;
  /** Default currency. */
  currency?: string;
  /** Agent reputation score (0.0 - 1.0). */
  reputation?: number;
  /** Agent availability score (0.0 - 1.0). */
  availability?: number;
  /** Whether settlement is required. */
  required?: boolean;
  /** Human-readable description. */
  description?: string;
}

/**
 * Build an AgentCard `capabilities.extensions[]` entry for A2A-SE.
 */
export function buildSettlementExtension(
  options: SettlementExtensionOptions,
): Record<string, unknown> {
  const params: Record<string, unknown> = {
    exchangeUrls: options.exchangeUrls,
    preferredExchange:
      options.preferredExchange ?? options.exchangeUrls[0],
    accountIds: options.accountIds,
    currency: options.currency ?? "ATE",
  };

  if (options.pricing) params["pricing"] = options.pricing;
  if (options.reputation !== undefined)
    params["reputation"] = options.reputation;
  if (options.availability !== undefined)
    params["availability"] = options.availability;

  return {
    uri: A2A_SE_EXTENSION_URI,
    description:
      options.description ??
      "Accepts token-based payment via A2A Settlement Exchange",
    required: options.required ?? false,
    params,
  };
}
